[DRAFT] Adding support for Qwen3 model

kurtis318 opened the pull request:
This is a request to add code changes for Qwen3 models. One outstanding test (tests/distributed/test_tp.py::test_attention_tp_fused) needs to be resolved.

ani300 commented on fms/modules/attention.py:
```diff
@@ -334,6 +336,7 @@ def __init__(
         self.emb_kq_per_head = emb_kq_per_head
         self.emb_v_per_head = emb_v_per_head
         self.use_bias = use_bias
+        self.use_norm = use_norm
```
no longer needed

kurtis318 replied:
Commenting out for now.

ani300 commented on fms/modules/attention.py:
```diff
@@ -570,26 +576,64 @@ def __init__(
         self.use_bias = use_bias
+        self.use_norm = use_norm
         self.fused = fused
         self.linear_config = linear_config
         self.scale_factor = scale_factor
 
-        self.in_proj: QKV = (FusedQKV if self.fused else UnfusedQKV)(
+        self.in_proj:QKV = None     # type:ignore
+        if self.fused:
+            self.in_proj = FusedQKV(
             self.emb_dim,
             self.nheads,
             self.kvheads,
             self.emb_kq_per_head,
             self.emb_v_per_head,
             self.use_bias,
-            linear_config=linear_config,
-        )
+            linear_config
+            )
+        else:
+            self.in_proj = UnfusedQKV(
+            self.emb_dim,
+            self.nheads,
+            self.kvheads,
+            self.emb_kq_per_head,
+            self.emb_v_per_head,
+            self.use_bias,
+            linear_config,
+            )
+        # self.in_proj: QKV = (FusedQKV if self.fused else )(
+        #     self.emb_dim,
+        #     self.nheads,
+        #     self.kvheads,
+        #     self.emb_kq_per_head,
+        #     self.emb_v_per_head,
+        #     self.use_bias,
+        #     self.use_norm,
+        #     linear_config=linear_config,
+        # )
```
this can all return to the original code

kurtis318 replied:
Commenting out to test. Will remove when runs.

kurtis318 replied:
test_qwen.py runs without error. Removed code

ani300 commented on tests/models/test_qwen.py:
```diff
@@ -0,0 +1,162 @@
+        return model.base_model
+
+
+class QwenGPTQFixtures(ModelFixtureMixin):
+    @pytest.fixture(scope="class", autouse=True)
+    def uninitialized_model(self):
+        """_summary_
+
+        Returns:
+            _type_: _description_
+        """
+        return Qwen(
+            src_vocab_size=384,
+            emb_dim=64,
+            norm_eps=1e-05,
+            nheads=32,
+            kvheads=8,
+            nlayers=2,
+            pad_id=0,
+            hidden_grow_factor=3.5,
+            multiple_of=2,
+            tie_heads=False,
+            activation_fn="swish",
+            sliding_window=4000,
+            rope_base=100_0000.0,
+            p_dropout=0.0,
+            max_expected_seq_len=4096,
+            linear_config={"linear_type": "gptq_cpu"},
+            fused_weights=True,
+        )
```
why is this defined twice?

ani300 commented on tests/models/test_qwen.py:
```diff
@@ -0,0 +1,162 @@
+            rope_base=100_0000.0,
+            p_dropout=0.0,
+            max_expected_seq_len=4096,
+            linear_config={"linear_type": "gptq_cpu"},
+            fused_weights=True,
+        )
+
+    def _maybe_get_initialized_parameter(self, key, parameter):
+        """_summary_
+
+        Args:
+            key (_type_): _description_
+            parameter (_type_): _description_
+
+        Returns:
+            _type_: _description_
+        """
+        if "qweight" in key:
+            return torch.randint(
+                low=0,
+                high=torch.iinfo(torch.int32).max,
+                size=parameter.shape,
+                dtype=torch.int32,
+            )
+        if "qzeros" in key:
+            return torch.ones(parameter.shape, dtype=torch.int32) * 8
+        if "g_idx" in key:
+            return parameter
+        return None
+
+
+@pytest.mark.autogptq
+class TestQwenGPTQ(
+    ModelConsistencyTestSuite, ModelCompileTestSuite, QwenGPTQFixtures
+):
+    # x is the main parameter for this model which is the input tensor
+    _get_signature_params = ["x"]
+
+    def test_model_unfused(self, model, signature):
+        pytest.skip("weight unfuse is not implemented for GPTQ")
```
unless we have gptq models for Qwen3, this test is not needed

kurtis318 replied:
commenting out for test

ani300 commented on fms/models/qwen.py:
```diff
@@ -0,0 +1,782 @@
+        cached_freqs: dict[Optional[torch.device], dict[int, torch.Tensor]],
+        max_seq_len_cached: dict[Optional[torch.device], int],
+    ):
+        """_summary_
+
+        Args:
+            cached_freqs (dict[Optional[torch.device], dict[int, torch.Tensor]]): _description_
+            max_seq_len_cached (dict[Optional[torch.device], int]): _description_
+        """
```
fill stubs

ani300 commented on fms/models/qwen.py:
```diff
@@ -0,0 +1,782 @@
+        config: QwenConfig,
+        distributed_strategy: DistributedStrategy = NoOpStrategy,
+    ):
+        """_summary_
+
+        Args:
+            config (QwenConfig): _description_
```
fill stubs or remove

ani300 commented on fms/models/qwen.py:
```diff
@@ -0,0 +1,782 @@
+    ):
+        """_summary_
+
+        Args:
+            x (_type_): _description_
+            mask (_type_, optional): _description_. Defaults to None.
+            position_ids (_type_, optional): _description_. Defaults to None.
+            past_key_value_state (_type_, optional): _description_. Defaults to None.
+            use_cache (bool, optional): _description_. Defaults to False.
+            is_causal_mask (bool, optional): _description_. Defaults to False.
+            attn_algorithm (_type_, optional): _description_. Defaults to None.
+
+        Returns:
+            _type_: _description_
```
fill or remove

ani300 commented on fms/models/qwen.py:
```diff
@@ -0,0 +1,782 @@
+        if past_key_value_states is None or len(past_key_value_states) == 0:
+            past_key_value_states = [None for _ in range(len(self.layers))]
+
+        qlen = x_in.size(1)
+        klen = x_in.size(1)
+
+        # if we are using the cache, the key length needs to be extended with the past keys length
+        if use_cache and past_key_value_states[0] is not None:
+            klen += past_key_value_states[0][0].size(-2)
+
+        # if mask is none, we need to specify causal mask
+        if mask is None:
+            # we are caching and can assume all 1s in the mask
+            if use_cache and klen != 1 and qlen == 1:
+                # b x h x qlen x kvlen
+                is_causal_mask = False
+            else:
+                is_causal_mask = True
+        else:
+            is_causal_mask = False
```
I think all this logic can be cleaned up, look at granite or llama in the latest fms

ani300 reviewed (commented):
Left a bunch of comments to be addressed, I'll take a look again after they are addressed.

kurtis318 commented on fms/modules/attention.py:
```diff
@@ -323,6 +324,7 @@ def __init__(
         emb_kq_per_head: int,
         emb_v_per_head: int,
         use_bias: bool,
+        use_norm: bool,
```
removed

kurtis318 commented on fms/modules/attention.py:
```diff
@@ -516,7 +519,8 @@ def forward(
         else:
             raise ValueError("q, k, and v must be the same or k and v must be None")
-        return self.qkv_fused(qkv).split(self.splits, dim=-1)
+        queries, keys, values = self.qkv_fused(qkv).split(self.splits, dim=-1)
+        return queries, keys, values
```
test_qwen.py runs without error. Removed code

kurtis318 commented on tests/models/test_qwen.py:
```diff
@@ -0,0 +1,162 @@
+            tie_heads=False,
+            activation_fn="swish",
+            sliding_window=4000,
+            rope_base=100_0000.0,
```
using 1000000.0 for test

kurtis318 commented:
Update on next steps.

Next steps
1. Fix unexpected keyword argument 'attn_name'?   IN-PROGRESS
2. Fix RuntimeError: GraphLoader CompileGraph failed: compile_graph failed? NOT STARTED
3. Does [WARNING] Keys from checkpoint (adapted to FMS) not copied into model need to be fixed?
  - The keys from the checkpoint in question end in `attn.in_proj.k_norm.weight` and `attn.in_proj.q_norm.weight`
  - These are being mapped by `qwen.py` function `_hf_to_fms_names()`. 
  - The weight name to  `_hf_to_fms_names()` is `self_attn.k_norm.weight` and `self_attn.q_norm.weigh`
  - Question: Can someone confirm if these two weights are indeed needed? Since they were mapped I assume they are needed.

kurtis318 commented:
#### Next steps

Completed:
1. Fixed `unexpected keyword argument 'attn_name'`?  Solution:  use [Small examples of using Foundation Model Stack (FMS) on AIU hardware](https://github.com/foundation-model-stack/aiu-fms-testing-utils/tree/main/examples#small-examples-of-using-foundation-model-stack-fms-on-aiu-hardware)
2. Fix `RuntimeError: GraphLoader CompileGraph failed: compile_graph failed`? use [Small examples of using Foundation Model Stack (FMS) on AIU hardware](https://github.com/foundation-model-stack/aiu-fms-testing-utils/tree/main/examples#small-examples-of-using-foundation-model-stack-fms-on-aiu-hardware)

Next steps
1. Determine why inference results are not correct.
3. Does  `[WARNING] Keys from checkpoint (adapted to FMS) not copied into model` need to be fixed?

kurtis318 commented on fms/models/qwen.py:
```diff
@@ -0,0 +1,782 @@
+            self.dropout = nn.Dropout(self.config.p_dropout)
+
+    def reset_parameters(self):
+        """_summary_
+        """
```
removed

kurtis318 commented:
Hear is a list of issues Kauoutar found with Qwen and discussed in the AIU Model Enablement  Squad meeting today.
Kauout scheduled 1 hr meeting to discuss this tomorrow (Oct 2).  Just getting a head start.

|#|Summary|Status|Detail|
|---|---|---|---|
|01|Wrong architecture mapping|working|FMS maps Qwen3ForCausalLM to "qwen" instead of "qwen3", so it loads the wrong adapters and skips many weights.|
|02|Incomplete config translation|working|Key parameters from HF (e.g., vocab size, number of layers/heads, head_dim, attention bias, rope_scaling) are missing, leading to shape and math drift.|
|03|Mismatch in MLP & activation|?|HF uses SwiGLU with explicit intermediate_size, while FMS uses a grow-factor heuristic with "swish", causing hidden size inconsistencies.|
|04|RoPE scaling ignored|?|HF supports advanced rope scaling variants (e.g., Llama-3 style). FMS ignores this, breaking positional encodings.|
|05|Weight-naming misalignment|next|HF’s q_proj/k_proj/v_proj and gate/up/down_proj don’t map to FMS’s fused qkv and wg1/w2 naming, so weights are skipped instead of remapped.|
|06|Cache & forward semantics differ|?|HF supports cache_position and DynamicCache. FMS forward path doesn’t align, causing mismatched outputs during generation.|
|07|Test defaults hide bugs|?|Unit tests use tiny configs with non-HF defaults, which mask real mismatches.|

- `01: Wrong architecture mapping`
	- The classes in `qwen.py` are `QwenConfig`, `QwenBlock`,  `QwenHeadless` and `Qwen`
	- I plan to make the following changes
		-  rename `qwen.py` to`qwen3.py`
		-  rename `QwenConfig` to `QwenConfig3`
		-  rename `QwenBlock` to `QwenBlock3`
		-  rename `QwenHeadless` to `QwenHeadless3`
		-  rename `Qwen` to `Qwen3`
		-  change all references to the 4 old class names to the new names
		-  The following lines in `qwen` need the letter three "3" added
		```python
		41:    architecture = "qwen"
		57:    "Qwen3ForCausalLM"
		70:  "model_type": "qwen3",
		88:    "Qwen3ForCausalLM"
		452:    def from_config(cls, config: QwenConfig) -> "Qwen":
		505:_ARCHITECTURE_NAME = "qwen"
		515:models.register_model(_ARCHITECTURE_NAME, "1.7b", _qwen_factory_factory(_1_7b_config))
		572:            "GPTQ HF Qwen checkpoints cannot be loaded into a model with fused weights"
		579:    _ARCHITECTURE_NAME, "hf_gptq_fusion_check", _hf_gptq_qwen_check
		```
	- Will this fix the architecture mapping issue?
- `02: Incomplete config translation`
	- I added all the settings (except architecture) from the `config.json` into the `QwenConfig` class in `qwen.py`
- `05: Weight-naming misalignment`
	- I added debug code debug code in  function `_hf_to_fms_names()` found in `qwen.py` to keep track of ALL input and output mappings
	- I removed this debug code in the repo but will add it back in.
	- I do not recall seeing `gate/up/down_proj` mappings.

kaoutar55 commented on fms/utils/serialization.py:
```diff
@@ -524,7 +528,7 @@ def load_state_dict_into_model(
             f"not copied into model: {unused_keys}"
-        )
+            )
 
```
The current loader relies on skips (KWR_SKIP) instead of adapting HF weight names. Please add a Qwen3 adapter in fms/utils/serialization.py that:
Concats q_proj/k_proj/v_proj → attn.in_proj.qkv_fused
Maps o_proj → attn.dense
Concats gate_proj/up_proj → ff_sub_layer.wg1_fused
Maps down_proj → ff_sub_layer.w2
Maps input/post norms to ln / ff_ln, final norm → dec_norm, and lm_head → head
After this, remove KWR_SKIP and assert zero missing keys in your tests.

In this context, "Map" means to rename and route a tensor from the HF checkpoint’s key to the key your FMS model expects. “Concat” means to combine multiple HF tensors into one FMS tensor because FMS fuses several linear layers into a single, bigger linear weight. Examples for concat: Attention QKV fusion and MLP gate+up fusion (SwiGLU)

kaoutar55 commented on fms/models/qwen3.py:
```diff
@@ -0,0 +1,701 @@
+ "architectures": [
+    "Qwen3ForCausalLM"
+  ],
+  "attention_bias": false,
```
Make sure your FMS Qwen3 model receives the HF flag:
In _infer_model_configuration, forward:
config_params["attention_bias"] = config.attention_bias
In fms/models/qwen3.py, store it:
self.attention_bias = bool(config.attention_bias)

Pass attention_bias from HF config into FMS config.
Build q/k/v/o (or fused qkv + o) linears with bias=self.attention_bias.
In the adapter, concat q/k/v biases if fused; map o-bias.
Add tests to ensure both bias-on and bias-off paths are correct and produce zero missing keys.

kaoutar55 reviewed (commented):
Update tests to import fms.models.qwen3, not fms.models.qwen.
Add a config-parity test comparing Qwen3Config with FMS-inferred config.
Add a checkpoint load test verifying no missing/unused keys.
Remove reliance on KWR_SKIP.

Overall I think we need the following:
(1) forward all Qwen3Config fields correctly, (2) add a proper state-dict adapter, (3) respect attention_bias and rope_scaling, and (4) fix tests. This will bring  hopefully the FMS Qwen3 implementation to parity with HF.

kaoutar55 commented on fms/models/qwen3.py:
```diff
@@ -0,0 +1,701 @@
+  + "num_hidden_layers": 28,            # nlayers (10)
+  + "num_key_value_heads": 8,           # kvheads (04)
+  + "rms_norm_eps": 1e-06,              # norm_eps (06)
+  "rope_scaling": null,
```
Qwen3Config has a rope_scaling = None field, but it’s never used

kaoutar55 commented on fms/models/qwen3.py:
```diff
@@ -0,0 +1,701 @@
+            padding_idx=self.config.pad_id,
+        )
+
+        self.rot_emb = RotaryEmbedding(
```
There’s no rope_scaling (variant) passed, only rope_base (theta).

kaoutar55 reviewed (commented):
RoPE scaling (the structured dict from HF, e.g. {"rope_type": "llama3", ...}) is not applied anywhere; angles will diverge from HF when scaling is present.

ani300 commented on fms/models/hf/utils.py:
```diff
@@ -227,6 +227,17 @@ def _infer_model_configuration(
+        config_params["p_dropout"] = config.attention_dropout
+        config_params["norm_eps"] = config.rms_norm_eps
+        config_params["rope_base"] = config.rope_theta
+        config_params["sliding_window"] = config.sliding_window
```
the first 3 are already forwarded at the end of this function, the others need forwarding

ani300 commented on fms/models/hf/utils.py:
add the model name here too

ani300 commented on fms/models/qwen3.py:
```diff
@@ -0,0 +1,701 @@
+    p_dropout: float = 0.0
+    activation_fn: str = "silu"
+    emb_dim: int = 2048
+    max_expected_seq_len: int = 40960
+    kvheads: int = 8
+    norm_eps: float = 1e-6
+    rope_base: float = 100_0000.0  # Same as rope_theta
+    # Extra?
+    bos_token_id:int = 151643
+    multiple_of: int = 256  # borrowed from llama
+    sliding_window: int = 4000
+    fused_weights: bool = True  # FMS Specific -- For CPU/GPU = T, AIU = F
+    pad_id: int = -1  # borrowed from granite, we do need it
+    linear_config: Optional[Mapping[str, Any]] = None  # To support quantization
+    # From config.json just in case?
+    attention_bias: bool = False
+    attention_dropout: float = 0.0
+    # bos_token_id: int = 151643
+    eos_token_id: int = 151645
+    head_dim:int = 128
+    hidden_act:str = "silu"
+    hidden_size: int = 2048
+    initializer_range: float = 0.02
+    intermediate_size:int = 6144
+    max_position_embeddings:int = 40960
+    max_window_layers:int = 28
+    model_type:str = "qwen3"
+    num_attention_heads:int = 16
+    num_hidden_layers:int = 28
+    num_key_value_heads:int = 8
+    rms_norm_eps:float = 1e-06
+    rope_scaling =  None
+    rope_theta:int = 1000000
+    sliding_window = 0  # not sure what to put here
+    tie_word_embeddings:bool = True
+    torch_dtype: str = "bfloat16"
+    transformers_version:str = "4.51.0"
+    use_cache:bool = True
+    use_sliding_window = False
+    vocab_size: int = 15193
```
you are repeating many parameters with different names, please carefully map the HF ones to FMS and not more, you should have at most max(FMS params, HF params), not a sum of both

ani300 reviewed (commented):
this will need to be rebased on top of the current main of FMS, there's a few APIs that have changed since then

kaoutar55 reviewed (commented):
One critical blocker remains: The weight adapter is not implemented. This means:

HuggingFace checkpoint weights are not being transformed to FMS format
Model is either using random weights or incorrectly mapped weights
Output will be completely wrong

The test using KWR_SKIP is essentially saying "skip weight loading validation because we know it's broken."
To fix the incorrect output, you MUST implement the weight adapter in serialization.py following the pattern used by LLaMA

flaviabeo reviewed (changes requested):
Hello! I tested a bit the PR and have some comments;

This is the simple code I used to test the inference:

```python
import torch

from fms.models import get_model
from fms.utils.generation import generate

from transformers import AutoTokenizer

qwen = get_model("hf_pretrained", "Qwen/Qwen3-8B")

tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen3-8B")
tokens = tokenizer.encode("Provide a list of instructions for preparing chicken soup.", return_tensors="pt")
result = generate(qwen, tokens, temperature=1.0, top_k=50, max_new_tokens=100, do_sample=False, use_cache=True)
```

kurtis318 commented on tests/models/test_qwen.py:
```diff
@@ -0,0 +1,161 @@
+import pytest
+import torch
+
+from fms.models.qwen import Qwen, QwenConfig, QwenHeadless
```
applied fix
